### PR TITLE
[VL-53] Get detail of notification

### PR DESCRIPTION
**Pull request by kin0992**
Get details about an accepted notification.

#### List of Changes
- Add schemas to the OpenAPI specification

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kin0992** [comment]
@gunzip do you have any idea why, if you run the `npm run generate` command, the file `src/generated/FullSentNotification.ts` has an error?

```
export const FullSentNotification = t.intersection(
  [, FullSentNotification2],
  "FullSentNotification"
);
```

**Comment by gunzip** [comment]
maybe because of the allOf? try extracting the `type: object` into its own definition and making it a $ref.

**Comment by kin0992** [comment]
> maybe because of the allOf? try extracting the `type: object` into its own definition and making it a $ref.

I tried that, but it didn't work and, for the records, the `NotificationDocument` schema has the `allOf` with an external reference using the `$ref` and a type declared within the `allOf`.

It looks like that at some point the generator goes crazy with external references `$ref: 'https://...`, but not with all the external references (e.g. the `PreLoadRequest` schema).

**Comment by kin0992** [comment]
I pushed a solution that compiles, but I had to import locally the some schemas

File: src/domain/GetNotificationDetailRepository.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import { Response, UnauthorizedMessageBody } from './types';
+import { Repository } from './Repository';
+
+export type GetNotificationDetailRecord = {
+  type: 'GetNotificationDetailRecord';
+  input: { apiKey: ApiKey; iun: Iun };
+  output: Response<200, FullSentNotification> | Response<403, UnauthorizedMessageBody> | Response<404>;
+};
+
+export type GetNotificationDetailRepository = Repository<GetNotificationDetailRecord>;
```
Comment (datalek) [inline src/domain/GetNotificationDetailRepository.ts:13]:
```suggestion
export type GetNotificationDetailRecordRepository = Repository<GetNotificationDetailRecord>;
```

Comment (kin0992) [reply src/domain/GetNotificationDetailRepository.ts]: Done!

File: src/domain/Notification.ts [hunk 2]
```diff
@@ -3,34 +3,40 @@ import * as M from 'fp-ts/Monoid';
...
       findNotificationRequestRecord.output.statusCode === 200
         ? O.some(findNotificationRequestRecord.output.returned)
         : O.none,
       O.filter((e) => e.notificationRequestId === notificationRequest.notificationRequestId),
       O.filterMap((e) => O.fromNullable(e.iun)),
-      O.map((iun) => ({ ...notificationRequest, iun }))
+      O.map((iun) => ({
+        ...notificationRequest,
+        ...makeFullSentNotification(senderPaId)(sentAt)(notificationRequest)(iun),
+      }))
```
Comment (datalek) [inline src/domain/Notification.ts:26]:
```suggestion
      O.map(makeFullSentNotification(senderPaId)(sentAt)(notificationRequest)),
      O.map((full) => ({ ...notificationRequest, ...full }))
```

File: src/domain/Notification.ts [hunk 3]
```diff
@@ -3,34 +3,40 @@ import * as M from 'fp-ts/Monoid';
...
   (consumeEventStreamRecord: ConsumeEventStreamRecord): O.Option<Notification> =>
     pipe(
       getProgressResponseList([consumeEventStreamRecord]),
       RA.findLast((e) => e.notificationRequestId === notificationRequest.notificationRequestId),
       O.filterMap((e) => O.fromNullable(e.iun)),
-      O.map((iun) => ({ ...notificationRequest, iun }))
+      O.map((iun) => ({
+        ...notificationRequest,
+        ...makeFullSentNotification(senderPaId)(sentAt)(notificationRequest)(iun),
+      }))
```
Comment (datalek) [inline src/domain/Notification.ts:39]:
```suggestion
      O.map(makeFullSentNotification(senderPaId)(sentAt)(notificationRequest)),
      O.map((full) => ({ ...notificationRequest, ...full }))
```

File: src/domain/__tests__/data.ts [hunk 4]
```diff
@@ -192,3 +191,8 @@ export const consumeEventStreamRecordDelivered = {
     })),
   },
 };
+
+export const acceptedNotification = makeFullSentNotification(aSenderPaId)(aDate)({
+  ...newNotificationRequest,
+  notificationRequestId: notificationId.valid,
+})(aIun.valid);
```
Comment (datalek) [inline src/domain/__tests__/data.ts:198]: [Optional]: I suggest the following change to improve the consistency with other tests, in addition in the future we may require `GetNotificationDetailRecord` data :)

```suggestion
// GetNotificationDetailRecord /////////////////////////////////////////////////

const acceptedNotification = makeFullSentNotification(aSenderPaId)(aDate)({
  ...newNotificationRequest,
  notificationRequestId: notificationId.valid,
})(aIun.valid);

export const getNotificationDetailRecordAccepted: GetNotificationDetailRecord = {
  type: 'GetNotificationDetailRecord',
  input: { apiKey: apiKey.valid, iun: aIun.valid },
  output: { statusCode: 200, returned: acceptedNotification }
};
```

File: src/useCases/__tests__/GetNotificationDetailUseCase.test.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      inMemory.makeRepository(logger)<CheckNotificationStatusRecord>([
+        data.checkNotificationStatusRecord,
+        data.checkNotificationStatusRecordAccepted,
+      ]),
+      inMemory.makeRepository(logger)<ConsumeEventStreamRecord>([]),
+      () => data.aIun.valid,
+      () => data.aDate
+    );
+
+    const expected = E.right({ statusCode: 200, returned: data.acceptedNotification });
```
Comment (datalek) [inline src/useCases/__tests__/GetNotificationDetailUseCase.test.ts:49]: [Attention!]: This suggestion requires the suggestion on [`src/domain/__tests__/data.ts`](https://github.com/pagopa/pn-local-emulator/pull/36/files#diff-871bd52c41bb599eb34134454ce218f9dec3881eb954088765176436c19b99e8) file

```suggestion
    const expected = E.right({ statusCode: 200, returned: data.getNotificationDetailRecordAccepted.output });
```

Comment (kin0992) [reply src/useCases/__tests__/GetNotificationDetailUseCase.test.ts]: Done!

File: src/domain/__tests__/data.ts (hunk 4 above)
Comment (kin0992) [reply src/domain/__tests__/data.ts]: Done!

File: src/useCases/__tests__/ConsumeEventStreamUseCase.test.ts [hunk 6]
```diff
@@ -11,12 +11,14 @@ import { ConsumeEventStreamRecord } from '../../domain/ConsumeEventStreamRecordR
 import { CheckNotificationStatusRecord } from '../../domain/CheckNotificationStatusRepository';
 
 const numberOfWaitingBeforeComplete = 2;
+const senderPaId = 'aSenderPaId';
```
[the remark this answers is not in the document]
Comment (kin0992) [reply src/useCases/__tests__/ConsumeEventStreamUseCase.test.ts]: Done!

File: src/useCases/__tests__/CheckNotificationStatusUseCase.test.ts [hunk 7]
```diff
@@ -9,11 +9,13 @@ import { ConsumeEventStreamRecord } from '../../domain/ConsumeEventStreamRecordR
 
 const logger = makeLogger();
 const numberOfWaitingBeforeComplete = 2;
+const senderPaId = 'aSenderPaId';
```
[the remark this answers is not in the document]
Comment (kin0992) [reply src/useCases/__tests__/CheckNotificationStatusUseCase.test.ts]: Done!

**Review by datalek — APPROVED** [review]
🐴

**Pull request merged**